examples/stein_bnn.py

```diff
@@ -0,0 +1,177 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+"""
+Example: Bayesian Neural Network with SteinVI
+=============================================
+We demonstrate how to use SteinVI to predict housing prices using a BNN for the Boston Housing prices dataset
+from the UCI regression benchmarks.
+"""
+
+import argparse
+from collections import namedtuple
+import datetime
+from functools import partial
+from time import time
+
+from sklearn.model_selection import train_test_split
+
+from jax import random
+import jax.numpy as jnp
+
+import numpyro
+from numpyro import handlers
+from numpyro.contrib.einstein import RBFKernel, SteinVI
+from numpyro.distributions import Gamma, Normal
+from numpyro.examples.datasets import BOSTON_HOUSING, load_dataset
+from numpyro.infer import Predictive, Trace_ELBO, init_to_uniform
+from numpyro.infer.autoguide import AutoDelta
+from numpyro.optim import Adagrad
+
+DataState = namedtuple("data", ["xtr", "xte", "ytr", "yte"])
+
+
+def load_data() -> DataState:
+    _, fetch = load_dataset(BOSTON_HOUSING, shuffle=False)
+    x, y = fetch()
+    xtr, xte, ytr, yte = train_test_split(x, y, train_size=0.90)
+
+    return DataState(*map(partial(jnp.array, dtype=float), (xtr, xte, ytr, yte)))
+
+
+def normalize(val, mean=None, std=None):
+    """Normalize data to zero mean, unit variance"""
+    if mean is None and std is None:
+        # Only use training data to estimate mean and std.
+        std = jnp.std(val, 0, keepdims=True)
+        std = jnp.where(std == 0, 1.0, std)
+        mean = jnp.mean(val, 0, keepdims=True)
+    return (val - mean) / std, mean, std
+
+
+def model(x, y=None, hidden_dim=50, subsample_size=100):
+    """BNN described in section 5 of [1].
+
+    **References:**
+        1. *Stein variational gradient descent: A general purpose bayesian inference algorithm*
+        Qiang Liu and Dilin Wang (2016).
+    """
+
+    prec_nn = numpyro.sample(
+        "prec_nn", Gamma(1.0, 0.1)
+    )  # hyper prior for precision of nn weights and biases
+
+    n, m = x.shape
+
+    with numpyro.plate("l1_hidden", hidden_dim):
+        b1 = numpyro.sample(  # prior l1 bias term
+            "nn_b1",
+            Normal(
+                0.0,
+                1.0 / prec_nn,
+            ),
+        )
+        assert b1.shape == (hidden_dim,)
+
+        with numpyro.plate("l1_feat", m):
```

Comment:
It is a good practice to add dim to plate statements

Reply:
It makes sense; I've updated with dim params set.